bin/ch-grow.py.in

```diff
@@ -269,32 +271,68 @@ class Copy(Instruction):
       return "%s -> %s" % (self.srcs, repr(self.dst))
 
    def execute_(self):
-      # The Dockerfile spec for COPY is complex and messy. This implementation
-      # is not conforming but hopefully comes close. Known nonconformance:
-      #
-      # 1. We use the Python glob semantics instead of the Go ones.
+      # The Dockerfile specification for COPY is complex, messy, inexplicably
+      # different from cp(1), and incomplete. We try to be bug-compatible with
+      # Docker but probably are not 100%. See the FAQ.
       #
-      # 2. With "docker build" COPY seems to follow cp(1) semantics if the
-      #    destination is an existing directory: sources are copied to within
-      #    that directory regardless of whether it has trailing slash. This
-      #    contrasts with the spec's implication that this only happens if a
-      #    trailing slash is present; otherwise, the destination is replaced.
-      #
-      # 3. Spec does not say what to do with symlinks. If the source is a
-      #    symlink to a file, we copy the link target to a regular file;
-      #    otherwise, we copy symlinks themselves.
-      srcs = itertools.chain.from_iterable(glob.glob(cli.context + "/" + i)
-                                           for i in self.srcs)
+      # None of this abstracted into a general copy function because the
+      # semantics are really weird, so I don't want people calling it except
+      # from here.
+      srcs = list()
+      for src in self.srcs:
+         if (os.path.normpath(src).startswith("..")):
+            FATAL("can't COPY: %s climbs outside context" % src)
+         for i in glob.glob(cli.context + "/" + src):
+            srcs.append(i)
+      if (len(srcs) == 0):
+         FATAL("can't COPY: no sources exist")
       dst = dir_image(cli.tag) + "/"
       if (not self.dst.startswith("/")):
          dst += state.workdir + "/"
       dst += self.dst
-      if (dst.endswith("/")):
-         dst = dst[:-1]
-         mkdirs(dir_image(cli.tag) + dst)
+      if (dst.endswith("/") or len(srcs) > 1 or os.path.isdir(srcs[0])):
+         # Create destination directory.
+         if (dst.endswith("/")):
+            dst = dst[:-1]
+         if (os.path.exists(dst) and not os.path.isdir(dst)):
+            FATAL("can't COPY: %s exists but is not a directory" % dst)
+         ch.mkdirs(dst)
       for src in srcs:
-         cp(src, dst)
-
+         # Check for symlinks to outside context.
+         src_real = os.path.realpath(src)
+         context_real = os.path.realpath(cli.context)
+         if (not os.path.commonpath([src_real, context_real]) \
+                 .startswith(context_real)):
+            FATAL("can't COPY: %s uses symlinks to get outside context" % src)
```

Comment:
We should consider stating that symlinks can't go outside the context in the documentation.

Reply:
OK, I'll think about it. I believe Docker's Dockerfile documentation talks about how you can only use files within the context directory.